contracts/ibc-callbacks/src/contract.rs

```diff
@@ -0,0 +1,174 @@
+use cosmwasm_std::{
+    entry_point, to_json_binary, to_json_string, Addr, Binary, Deps, DepsMut, Empty, Env,
+    IbcBasicResponse, IbcMsg, IbcPacketAckMsg, IbcPacketTimeoutMsg, IbcTimeout, MessageInfo,
+    Response, StdError, StdResult, Uint64,
+};
+use schemars::JsonSchema;
+use serde::{Deserialize, Serialize};
+
+use crate::msg::{ExecuteMsg, QueryMsg};
+use crate::state::{load_stats, save_stats, CallbackStats};
+
+#[entry_point]
+pub fn instantiate(
+    deps: DepsMut,
+    _env: Env,
+    _info: MessageInfo,
+    _msg: Empty,
+) -> StdResult<Response> {
+    // initialize counts
+    let counts = CallbackStats::default();
+    save_stats(deps.storage, &counts)?;
+
+    Ok(Response::new().add_attribute("action", "instantiate"))
+}
+
+/// Sends an `IbcMsg::Transfer` to the given `to_address` on the given `channel_id`.
+#[entry_point]
+pub fn execute(
+    _deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    msg: ExecuteMsg,
+) -> StdResult<Response> {
+    let transfer_msg = match &*info.funds {
+        [coin] if !coin.amount.is_zero() => IbcMsg::Transfer {
+            to_address: msg.to_address,
+            timeout: IbcTimeout::with_timestamp(env.block.time.plus_seconds(5)), // TODO: how much time?
+            channel_id: msg.channel_id.clone(),
+            amount: coin.clone(),
+            memo: Some(to_json_string(&IbcCallbackData::source(IbcSrcCallback {
+                address: env.contract.address.clone(),
+                gas_limit: None,
```

Comment:
Haha, so the contract says: _sure YOLO_ and then the relayer actually has to pay the gas the callback consumes? Is this how it works?

Reply:
Yes, the relayer pays the gas for the callback.
However, the callback middleware has a global limit per callback (but that will probably be very large by nature) and this `gas_limit` here just allows the user to further lower that (providing a smaller upper bound on gas cost for the relayer). The relayer can also configure their own limit and if that is reached, the tx is rolled back and another relayer can pick it up. I think if the global or user limit is reached, it just fails.